### DISCUSSION TITLE
使用 LSTM_AL 的說明及幾個問題

### DISCUSSION THREAD

**Discussion by 4hfly** [4hfly/Associated_Learning#10] [Ideas]
# Architecture
see [model.py](https://github.com/4hfly/Associated_Learning/blob/dev/model.py)
## _class_ `LSTM_AL(*args, **kwargs)`
* 使用 LSTM 的參數和 [Pytorch](https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html#torch.nn.LSTM) 版本的全數相同，唯一差別在需加入 output_size。
  ### parameters
  * input_size: Dimension of x or embedding size.
  * output_size: Dimension of y or embedding size.
  * hidden_size
  * num_layers: Default `1`.
  * batch_first: Default `False`.
  * batch_size: Default `1`.
  * dropout: Default `0`.
  * bidirectional: Default `False`.
## Hyperparameters
* 參數有[獨立出來](https://github.com/4hfly/Associated_Learning/blob/dcdfdf3b578378f048f041ff997936937eeaf5cd/model.py#L10)，未來可使用`.json`控制。
* 如果要使用其他參數如`bias`等，必須先到 [ALComponent()](https://github.com/4hfly/Associated_Learning/blob/dcdfdf3b578378f048f041ff997936937eeaf5cd/model.py#L64) 中修改。
## _class_ `ALComponent()`
* LSTM_AL 繼承這個 class，往後要新增的 nn 也都會繼承它。
* 缺點是 function f, b, g 寫起來可讀性較差，這部分可再討論。

# 幾個想討論的問題

![image](https://user-images.githubusercontent.com/4437259/123118296-172cb480-d475-11eb-8672-bf4903319f0b.png)

* LSTM_AL 中我沒有使用 bridge function，原因在於 f, g 使用的 hyperparameter 是一樣的，所以 hidden size 也相同。假如 f, g 不相同的話，那會需要兩邊都使用 bridge function？
* LSTM 的 output 會是：`output, (hidden_state, cell_state)`，因此計算 loss 的方式為
    ```
    output_x, (h_x, c_x) = LSTM_AL(x)
    output_y, (h_y, c_y) = LSTM_AL(y)
    loss = MSE(h_x, h_y)
    ```
    這部分應該沒問題吧？
* 如果要 stacked LSTM_AL 的話，有個小問題是，因為 y 會經過一個 decoder，所以 output size 會跟原來一樣。但 x 卻會是 hidden size 的大小，雖然這算很正常的狀況，不過我還是想確認一下是否有需要特別處理。

以上，之後會再補上 GRU_AL。

**Comment by Hibb-bb** [comment]
不確定ｙ經過的decoder是指哪個部分，是說AL的auto encoder 還是seq2seq的decoder? 如果是AL的話，你應該是指另外一項objective吧，我覺得好像可以用output_y and input_y 去算mse?因為size相同，不過考慮到language space 的特性，這邊先提出可以嘗試cosine similarity loss

**Reply by 4hfly** [reply]
> 我覺得好像可以用output_y and input_y 去算mse?

沒錯，我是在說 auto encoder 的部分，而現在就是照老師的方法先用 MSE。我回頭檢查過一遍之後，覺得第三個問題應該多餘了，是我沒想清楚，已經把正確的 code 更新上來了。
